controllers/eventhandlers/mapper.go

```diff
@@ -63,49 +66,78 @@ func (r *resourceMapper) EndpointsToService(ctx context.Context, ep *corev1.Endp
-			"policyName", policyName, "targetRef", targetRef)
-		return nil
+func (r *resourceMapper) VpcAssociationPolicyToGateway(ctx context.Context, vap *v1alpha1.VpcAssociationPolicy) *gateway_api.Gateway {
+	return policyToTargetRefObj(r, ctx, vap, &gateway_api.Gateway{})
+}
+
+func policyToTargetRefObj[T client.Object](r *resourceMapper, ctx context.Context, policy core.Policy, retObj T) T {
+	null := *new(T)
+	if policy == nil {
+		return null
 	}
+	policyNamespacedName := policy.GetNamespacedName()
 
-	svcName := types.NamespacedName{
-		Namespace: namespace,
+	targetRef := policy.GetTargetRef()
+	if targetRef == nil {
+		r.log.Infow("Policy does not have targetRef, skipping",
+			"policyName", policyNamespacedName)
+		return null
+	}
+	expectedGroup, expectedKind, err := k8sResourceTypeToGroupNameAndKindName(retObj)
+	if err != nil {
+		r.log.Errorw("Failed to get expected GroupVersion of targetRefObj",
+			"policyName", policyNamespacedName, "reason", err.Error())
+		return null
+	}
+
+	if targetRef.Group != gateway_api.Group(expectedGroup) || targetRef.Kind != gateway_api.Kind(expectedKind) {
+		r.log.Infow("Detected targetRef GroupVersion and expected retObj GroupVersion are different, skipping",
+			"policyName", policyNamespacedName,
+			"targetRef", targetRef,
+			"expectedGroup", expectedGroup,
+			"expectedKind", expectedKind)
+		return null
+	}
+	if targetRef.Namespace != nil && policyNamespacedName.Namespace != string(*targetRef.Namespace) {
+		r.log.Infow("Detected Policy and TargetRef namespace are different, skipping",
+			"policyName", policyNamespacedName, "targetRef", targetRef,
+			"policyNamespace", policyNamespacedName.Namespace)
+		return null
+	}
+
+	key := types.NamespacedName{
+		Namespace: policyNamespacedName.Namespace,
 		Name:      string(targetRef.Name),
 	}
-	svc := &corev1.Service{}
-	if err := r.client.Get(ctx, svcName, svc); err != nil {
+	if err := r.client.Get(ctx, key, retObj); err != nil {
 		if errors.IsNotFound(err) {
-			r.log.Debugw("TargetGroupPolicy is referring to non-existent service, skipping",
-				"policyName", policyName, "serviceName", svcName.String())
+			r.log.Debugw("Policy is referring to a non-existent targetRefObj, skipping",
+				"policyName", policyNamespacedName, "targetRef", targetRef)
 		} else {
 			// Still gracefully skipping the event but errors other than NotFound are bad sign.
 			r.log.Errorw("Failed to query targetRef of TargetGroupPolicy",
-				"policyName", policyName, "serviceName", svcName.String(), "reason", err.Error())
+				"policyName", policyNamespacedName, "targetRef", targetRef, "reason", err.Error())
 		}
-		return nil
+		return null
 	}
 	r.log.Debugw("TargetGroupPolicy change on Service detected",
-		"policyName", policyName, "serviceName", svcName.String())
+		"policyName", policyNamespacedName, "targetRef", targetRef)
 
-	return svc
+	return retObj
+}
+
+func k8sResourceTypeToGroupNameAndKindName(obj client.Object) (string, string, error) {
```

Comment:
nit: I would prefer GroupKind instead of GroupNameAndKindName, and perhaps use metav1.GroupKind to return object

Reply:
Changed in the new revision